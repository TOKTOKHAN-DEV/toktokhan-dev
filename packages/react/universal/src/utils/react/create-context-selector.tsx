import { ComponentType, ReactNode } from 'react'

import { createContext, useContextSelector } from 'use-context-selector'

import { PropsOf } from '../../types'

export type CreateContextSelectorReturn<T, P> = {
  useContext: <Selected>(selector: (value: T) => Selected) => Selected
  Provider: ({
    children,
    params,
  }: {
    children: ReactNode
    params?: P
  }) => JSX.Element
  withProvider: <C extends ComponentType<any>>(
    Component: C,
    params?: P,
  ) => (props: PropsOf<C>) => JSX.Element
}

/**
 * @category Utils/React
 * 커스텀 훅을 기반으로 컨텍스트와 관련된 유틸리티를 생성하는 함수입니다.
 * 이 함수는 주어진 훅을 컨텍스트로 감싸는 `Provider`, `useContext` 훅,
 * 그리고 컴포넌트를 컨텍스트로 감싸는 `withProvider` HOC를 반환합니다.
 *
 * @template T - 컨텍스트에서 사용할 데이터 타입
 * @template P - 훅의 파라미터 타입
 * @param useHook - 컨텍스트에서 사용할 커스텀 훅
 * @param initialProps - `useHook`에 전달될 초기 파라미터 (선택적)
 * @returns `{ useContext, Provider, withProvider }` - 생성된 컨텍스트 유틸리티들
 *
 * @example
 * ```tsx
 * // 커스텀 훅 정의
 * const useTimer = ({ timeLimit = 1000 }: { timeLimit?: number }) => {
 *   const [time, setTime] = useState(timeLimit);
 *   // 타이머 로직...
 *   return { time, start: () => { //타이머 시작 }};
 * };
 *
 * // createContextSelector로 컨텍스트 유틸리티 생성
 * const { Provider: TimerProvider, useContext: useTimerContext } = createContextSelector(useTimer);
 *
 * // 타이머를 표시하는 컴포넌트
 * const TimerDisplay = () => {
 *   const time = useTimerContext(ctx => ctx.time);
 *   return <div>Time: {time}</div>;
 * };
 *
 * // 방법 1. TimerProvider로 감싸기
 * const App = () => (
 *   <TimerProvider params={{timeLimit: 1000}}>
 *     <TimerDisplay />
 *   </TimerProvider>
 * );
 *
 * // 방법 2. withProvider로 컴포넌트 감싸기
 * const App = () => (
 *     <TimerDisplay />
 * );
 * export default withTimerProvider(App, { timeLimit: 1000 });
 * ```
 *
 */

export const createContextSelector = <T, P = undefined>(
  useHook: (param?: P) => T,
  initialProps?: P,
): CreateContextSelectorReturn<T, P> => {
  const context = createContext({} as T)

  const useContext = <Selected,>(selector: (value: T) => Selected) =>
    useContextSelector<T, Selected>(context, selector)

  const Provider = ({
    children,
    params,
  }: {
    children: ReactNode
    params?: P
  }) => {
    const value = useHook(params || initialProps)
    return <context.Provider value={value}>{children}</context.Provider>
  }

  const withProvider =
    <C extends ComponentType<any>>(Component: C, params?: P) =>
    // eslint-disable-next-line react/display-name
    (props: PropsOf<C>) => {
      return (
        <Provider params={params || initialProps}>
          <Component {...props} />
        </Provider>
      )
    }

  return { useContext, Provider, withProvider }
}
