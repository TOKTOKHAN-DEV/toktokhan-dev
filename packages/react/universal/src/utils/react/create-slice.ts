import { Immutable, produce } from 'immer'

import { ActionsByMap, ReducerMap } from './types/reducer'

type CreateReducerParams<S, R extends ReducerMap<S, any>> = {
  initialState: S
  reducers: R
}

const NOTHING: unique symbol = Symbol.for('immer-nothing')

type ValidRecipeReturnType<State> =
  | State
  | void
  | undefined
  | (State extends undefined ? typeof NOTHING : never)

/**
 * @param
 * @category Utils/React
 */
export const createSlice = <S, R extends ReducerMap<S, any>>({
  initialState,
  reducers,
}: CreateReducerParams<S, R>) => {
  const reducer = produce((state: S, action: ActionsByMap<S, R>) => {
    const reducer = reducers[action.type]
    return reducer(state, action.payload)
  })
  return { initialState, reducer } as {
    initialState: S
    reducer: void | S extends ValidRecipeReturnType<S> ?
      (base: Immutable<S>, action: ActionsByMap<S, R>) => S
    : never
  }
}
