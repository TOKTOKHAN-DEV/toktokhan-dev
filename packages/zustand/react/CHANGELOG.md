# @toktokhan-dev/zustand-react

## 0.0.1

### Patch Changes

- 4ab5fd0: add packages, fix type

  ### new package: zustand-react

  `@toktokhan-dev/zustand-react` 패키지는 `@toktokhan-dev/zustand-create-store-context` 와 `@toktokhan-dev/zustand-with-setter` 패키지에 있는 모든 모듈을 제공합니다.

  ### zustand-with-setter

  type 명이 수정되었습니다. `WithSetter` -> `IWithSetter`

- Updated dependencies [4ab5fd0]
  - @toktokhan-dev/zustand-with-setter@0.0.5
