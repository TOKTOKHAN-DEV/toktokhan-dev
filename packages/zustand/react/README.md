# @toktokhan-dev/zustand-react

`@toktokhan-dev/zustand-react` 패키지는 `@toktokhan-dev/zustand-create-store-context` 와 `@toktokhan-dev/zustand-with-setter` 패키지에 있는 모든 모듈을 제공합니다.

자세한 내용과 사용법은 [Tokdocs 공식 문서](https://toktokhan-dev-docs.vercel.app/docs/docs/zustand/Overview)에서 확인 할 수 있습니다.

## Installation

```
npm i @toktokhan-dev/zustand-react
```

## Npm packages

- [@toktokhan-dev/zustand-with-setter](https://www.npmjs.com/package/@toktokhan-dev/zustand-with-setter)
- [@toktokhan-dev/zustand-create-store-context](https://www.npmjs.com/package/@toktokhan-dev/zustand-create-store-context)

## Preview

### withSetter

`withSetter`를 사용하면 store 에 set 및 reset 함수를 추가하여, 간단한 상태변경은 함수를 따로 정의 하지 않아도 됩니다.

```ts
import { withSetter } from '@toktokhan-dev/zustand-react'

import { create } from 'zustand'

type Store = {
  count: number
  nested: { count: number; list?: string[] }
}

const useStore = create(
  withSetter<Store>(() => ({
    count: 0,
    nested: { count: 0 },
  })),
)

const set = useStore((store) => store.set)

set({ count: 5 })
set('count', 5)
set('nested.count', (prev) => prev + 1)
```

### createStoreContext

zustand 의 create 와 완전히 동일한 값을 받아 context 를 생성할 수 있습니다.

```tsx
import { createStoreContext } from '@toktokhan-dev/zustand-react'

type Store = {
  count: number
  setCount: (count: number) => void
}

const { Provider, useContext, withProvider } = createStoreContext<Store>(
  (set, get, store) => ({
    count: 0,
    setCount: (count: number) => set(() => ({ count })),
  }),
)
```
