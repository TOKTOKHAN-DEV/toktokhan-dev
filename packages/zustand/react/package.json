{
  "name": "@toktokhan-dev/zustand-react",
  "version": "0.0.1",
  "description": "",
  "type": "module",
  "main": "dist/index.js",
  "files": [
    "dist",
    "!dist/*.map"
  ],
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "require": "./dist/index.cjs",
      "import": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "rollup -c",
    "build:watch": "rollup -c --watch",
    "play": "pnpm build:watch & node --watch play/playground.js",
    "test": "jest",
    "api-extractor": "api-extractor run --local --verbose"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@toktokhan-dev/ts-config": "workspace:*"
  },
  "dependencies": {
    "@toktokhan-dev/zustand-create-store-context": "workspace:*",
    "@toktokhan-dev/zustand-with-setter": "workspace:*"
  }
}
