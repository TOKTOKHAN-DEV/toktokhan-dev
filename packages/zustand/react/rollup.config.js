import { setUpRollupByPackageJson } from '@toktokhan-dev/rollup-config/base.js'

const config = [
  ...setUpRollupByPackageJson({
    packageJsonPath: 'package.json',
    entry: 'src/index.ts',
    formats: ['cjs', 'es', 'dts'],
  }),
]

export default config
