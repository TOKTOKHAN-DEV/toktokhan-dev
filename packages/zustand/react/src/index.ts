export * from '@toktokhan-dev/zustand-create-store-context'
export * from '@toktokhan-dev/zustand-with-setter'

/**
 *
 * `@toktokhan-dev/zustand-create-store-context` 와 `@toktokhan-dev/zustand-create-store-context` 패키지에 있는 모든 모듈을 제공합니다..
 *
 * @packageDocumentation
 */
