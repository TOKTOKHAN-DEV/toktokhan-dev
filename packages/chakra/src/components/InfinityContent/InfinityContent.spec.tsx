/**
 * @jest-environment jsdom
 */
import { useRef } from 'react'

import '@testing-library/jest-dom'
import { render, screen, waitFor } from '@testing-library/react'

import { InfinityContent } from './InfinityContent'

jest.mock('@toktokhan-dev/react-web', () => ({
  useIntersectionObserver: jest.fn().mockImplementation(({ onVisible }) => {
    const targetRef = useRef<HTMLDivElement>(null)
    setTimeout(() => {
      onVisible()
    }, 1000)

    return { targetRef }
  }),
}))

describe('InfinityContent', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should render children when data is provided', () => {
    render(
      <InfinityContent
        hasMore={false}
        isFetching={false}
        onFetchMore={() => {}}
      >
        <div title={'@content'}>Content</div>
      </InfinityContent>,
    )

    expect(screen.getByTitle('@content')).toBeInTheDocument()
  })

  it('should show spinner when fetching data', () => {
    render(
      <InfinityContent
        hasMore={true}
        isFetching={true}
        onFetchMore={() => {}}
        spinner={<div title={'@spinner'}>Spinner</div>}
      >
        <div title={'@content'}>Content</div>
      </InfinityContent>,
    )

    expect(screen.getByTitle('@spinner')).toBeInTheDocument()
  })

  it('should call onFetchMore when onVisible called', async () => {
    const onFetchMore = jest.fn()

    render(
      <InfinityContent
        hasMore={true}
        isFetching={false}
        onFetchMore={onFetchMore}
      >
        <div title={'@content'}>Content</div>
      </InfinityContent>,
    )

    expect(onFetchMore).not.toHaveBeenCalled()

    await waitFor(
      () => {
        expect(onFetchMore).toHaveBeenCalled()
        expect(onFetchMore).toHaveBeenCalledTimes(1)
      },
      { timeout: 1500 },
    )
  })
})
