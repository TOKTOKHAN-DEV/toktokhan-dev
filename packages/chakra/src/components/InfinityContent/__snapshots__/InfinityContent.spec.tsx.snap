// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfinityContent should handle custom styles 1`] = `
<div>
  <div
    class="css-104ua2j"
  >
    <div
      title="@content"
    >
      Content
    </div>
    <div
      class="css-1r0rfp"
    />
  </div>
</div>
`;

exports[`InfinityContent should handle custom styles:  <div>
        <div
    +     aria-label="container"
          class="css-104ua2j"
        >
          <div
            title="@content"
          >
            Content
          </div>
          <div
    +       aria-label="flag"
            class="css-1r0rfp"
          />
        </div>
      </div> 1`] = `
<div>
  <div
    aria-label="container"
    class="css-104ua2j"
  >
    <div
      title="@content"
    >
      Content
    </div>
    <div
      aria-label="flag"
      class="css-1r0rfp"
    />
  </div>
</div>
`;
